---
title: ROSMAP Microglia Analysis
author: Effect of *APOE* Alleles on the Glial Transcriptome in Normal Aging and Alzheimerâ€™s Disease
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
    toc: yes
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

This script should be run first.

# Data Setup

```{r}
rm(list = ls())
# require(devtools)
# devtools::install_github('cttobin/ggthemr')
require(limma)
require(ComplexHeatmap)
require(data.table)
require(pheatmap)
require(ggplot2)
require(gridExtra)
require(ggthemr)
require(SNFtool)
require(circlize)
require(patchwork)
require(cowplot)
require(ggpubr)

ds = "ROSMAP"
dtype = "FPKM"
expSet = readRDS("ROSMAP-24.Rds")

# load expression data 
mData = expSet$mData
geneIds = rownames(mData)
samples = colnames(mData)
```


# Microglia Analysis

```{r}
celltype = "microglia"

# select microglia genes (n = 574) only
genes = readLines("Data/Microglia Genes.txt")

# select microglia genes that exist in ROSMAP
# after selection (n = 562)
annot = fread("Data/ENSEMBL GRCh38.p7.csv")
ensemblId_genes = unique(annot$EnsemblID[annot$GeneSymbol %in% genes])
ensemblIds = intersect(geneIds,ensemblId_genes)
annot2 = unique(annot[, .(EnsemblID, GeneSymbol, Description)]) # annotation with description
annot = unique(annot[, .(EnsemblID, GeneSymbol)])               # annotation without description
annot2 = annot2[EnsemblID %in% ensemblIds,]
annot = annot[EnsemblID %in% ensemblIds,]
length(ensemblIds)
```


# Pre-Processing

```{r}
# data selection
sel.gen = geneIds %in% ensemblIds 
mDataG = mData[sel.gen,] # for downstream analysis
mDataRawG = mDataG       # copy of raw data
geneIdsG = geneIds[sel.gen]
geneSymbols = annot$GeneSymbol[match(geneIdsG, annot$EnsemblID)]


# normalization
cat("Data Type: ", dtype, "\n")
mDataG = log2(mDataG + 0.05)
cat("Data Process: log2(FPKM + 0.05)\n")


# metadata
C = sapply(colnames(mData)[colnames(mData)!="Gene"], 
            function(x) expSet$strat[["CERAD"]][match(x, SubjectID)][["CERAD"]])
B = sapply(colnames(mData)[colnames(mData)!="Gene"], 
            function(x) expSet$strat[["Braak"]][match(x, SubjectID)][["Braak"]])
E = sapply(colnames(mData)[colnames(mData)!="Gene"], 
           function(x) expSet$strat[["APOE"]][match(x, SubjectID)][["APOE"]])
RawAPOE = sapply(colnames(mData)[colnames(mData)!="Gene"], 
           function(x) expSet$strat[["RawAPOE"]][match(x, SubjectID)][["RawAPOE"]])
```


# Run Models
* Model 1: `exp ~ APOE + CERAD | exp ~ APOE + Braak`
* Model 1.5: `exp ~ APOE + CERAD e2, e4 dosage model`
* Model 1.5.2: `exp ~ APOE + CERAD e2, e4 dosage model for C0 subjects only`
* Model 2 baseline: `C0E4-C0E3, C0E2-C0E3 | B1E4-B1E3, B1E2-B1E3`
* Model 3: `C3E4-C3E3, C3E2-C3E3 | B3E4-B3E3, B3E2-B3E3`
* Model 4: `C23E4-C23E3, C23E2-C23E3`
* Model 5: `C1E4-C1E3, C1E2-C1E3`

```{r}
source("Models.R")

# run Model 1
tT_CERAD = run_model1("CERAD", mDataG, C, E, annot, "ROSMAP", "microglia")
tT_Braak = run_model1("Braak", mDataG, B, E, annot, "ROSMAP", "microglia")

length(intersect(tT_Braak[E4vsE3_P.Value < 0.05]$GeneSymbol,tT_CERAD[E4vsE3_P.Value < 0.05]$GeneSymbol)) 

length(tT_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)  
length(tT_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)

length(tT_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC > 0,]$GeneSymbol)  
length(tT_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC < 0,]$GeneSymbol)  

length(tT_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)  
length(tT_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)  

# Table S4
# differentially expressed microglia genes across APOE groups for ROSMAP DLPFC and each of MSBB brain regions
TableS4 = annot[order(GeneSymbol), ]
TableS4[annot2, on = "EnsemblID", Description := i.Description]
TableS4[tT_CERAD, on="EnsemblID",  c("ROSMAP_E4vsE3_logFC", "ROSMAP_E4vsE3_P.Value", "ROSMAP_E4vsE3_adj.P.Val", "ROSMAP_E2vsE3_logFC", "ROSMAP_E2vsE3_P.Value", "ROSMAP_E2vsE3_adj.P.Val") := .(i.E4vsE3_logFC, i.E4vsE3_P.Value, i.E4vsE3_adj.P.Val,  i.E2vsE3_logFC, i.E2vsE3_P.Value, i.E2vsE3_adj.P.Val)]
fwrite(TableS4, "TableS4.csv")


# run Model 1.5 with e4 & e2 copies as numeric
E4Num = ifelse(RawAPOE == 44, 2, 
        ifelse(RawAPOE %in% c(34, 24), 1,
        ifelse(RawAPOE %in% c(22, 33, 23), 0, NA)))
E2Num = ifelse(RawAPOE == 22, 2, 
        ifelse(RawAPOE %in% c(23, 24), 1,
        ifelse(RawAPOE %in% c(44, 33, 34), 0, NA)))
tT_CERAD_ENum = run_model1.5("CERAD", mDataG, C, E4Num, E2Num, annot, "ROSMAP", "microglia")
tT1.sfE = topTable(tT_CERAD_ENum, sort.by="none", number=Inf, coef=2, confint=TRUE)
tT2.sfE = topTable(tT_CERAD_ENum, sort.by="none", number=Inf, coef=3, confint=TRUE)

E4Coeff = tT_CERAD_ENum$coefficients[,2]
E2Coeff = tT_CERAD_ENum$coefficients[,3]
all(names(E4Coeff) == tT1.sfE$EnsemblID)

E4.Genes = intersect(tT1.sfE[tT1.sfE$P.Value < 0.05 & E4Coeff > 0, "GeneSymbol"], tT_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
E2.Genes = intersect(tT2.sfE[tT2.sfE$P.Value < 0.05 & E2Coeff < 0, "GeneSymbol"], tT_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)

E4.Genes.2 = tT1.sfE[tT1.sfE$P.Value < 0.05 & E4Coeff > 0, "GeneSymbol"]
E2.Genes.2 = tT2.sfE[tT2.sfE$P.Value < 0.05 & E2Coeff < 0, "GeneSymbol"]
intersect(E4.Genes.2, E2.Genes.2)

# Table S5
# differentially expressed microglia genes across APOE groups for ROSMAP DLPFC and each of MSBB brain regions using a dosage model
setDT(tT1.sfE)
setDT(tT2.sfE)
TableS5 = annot[order(GeneSymbol), ]
TableS5 = TableS5[tT1.sfE, on = .(EnsemblID, GeneSymbol), c("ROSMAP_e4_dosage_logFC", "ROSMAP_e4_dosage_P.Value", "ROSMAP_e4_dosage_adj.P.Val") := .(i.logFC, i.P.Value, i.adj.P.Val)]
TableS5 = TableS5[tT2.sfE, on = .(EnsemblID, GeneSymbol), c("ROSMAP_e2_dosage_logFC", "ROSMAP_e2_dosage_P.Value", "ROSMAP_e2_dosage_adj.P.Val") := .(i.logFC, i.P.Value, i.adj.P.Val)]
fwrite(TableS5, "TableS5.csv")


# run Model 2
tT2_CERAD = run_model2("CERAD", mDataG, C, E, annot, "ROSMAP", "microglia") 
tT2_Braak = run_model2("Braak", mDataG, B, E, annot, "ROSMAP", "microglia")

nrow(tT2_CERAD[E4vsE3_P.Value < 0.05])
nrow(tT2_Braak[E4vsE3_P.Value < 0.05])

length(intersect(tT2_Braak[E4vsE3_P.Value < 0.05]$GeneSymbol,tT2_CERAD[E4vsE3_P.Value < 0.05]$GeneSymbol)) 

length(tT2_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)  
length(tT2_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)  

length(tT2_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC > 0,]$GeneSymbol)  
length(tT2_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC < 0,]$GeneSymbol)

length(tT2_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
length(tT2_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)


# run Model 1.5.2 with e4 & e2 copies as numeric C0 subjects only

# select C0 subjects only
cSubject = 0
# remove the 3 low-expressed genes: ENSG00000124091, ENSG00000252798, ENSG00000204345
mDataG1 = mDataG[!rownames(mDataG) %in% c("ENSG00000124091","ENSG00000252798","ENSG00000204345"),]
# select only the C0 subjects and remove the subject with APOE NA
mDataG1 = mDataG1[, which(C == cSubject)]
RawAPOE.C0 = RawAPOE[which(C == cSubject)]
E4Num.C0 = ifelse(RawAPOE.C0 == 44, 2, 
       ifelse(RawAPOE.C0 %in% c(34, 24), 1,
       ifelse(RawAPOE.C0 %in% c(22, 33, 23), 0, NA)))
E2Num.C0 = ifelse(RawAPOE.C0 == 22, 2, 
       ifelse(RawAPOE.C0 %in% c(23, 24), 1,
       ifelse(RawAPOE.C0 %in% c(44, 33, 34), 0, NA)))
tT_CERAD_ENum.C0 = run_model1.5.2(mDataG1, 
                                  E4Num.C0, E2Num.C0, annot, "ROSMAP", "microglia")
tT1.sfE.C0 = topTable(tT_CERAD_ENum.C0, sort.by="none", number=Inf, coef=2, confint=TRUE)
tT2.sfE.C0 = topTable(tT_CERAD_ENum.C0, sort.by="none", number=Inf, coef=3, confint=TRUE)
nrow(tT1.sfE.C0[tT1.sfE.C0$P.Value < 0.05, ])  
nrow(tT2.sfE.C0[tT2.sfE.C0$P.Value < 0.05, ])

E4Coeff.C0 = tT_CERAD_ENum.C0$coefficients[,2]
E2Coeff.C0 = tT_CERAD_ENum.C0$coefficients[,3]
all(names(E4Coeff.C0) == tT1.sfE.C0$EnsemblID)

E4.Genes.C0 = intersect(tT1.sfE.C0[tT1.sfE.C0$P.Value < 0.05 & E4Coeff.C0 > 0, "GeneSymbol"], tT2_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
E2.Genes.C0 = intersect(tT2.sfE.C0[tT2.sfE.C0$P.Value < 0.05 & E2Coeff.C0 < 0, "GeneSymbol"], tT2_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)

E4.Genes.C0.1 = tT1.sfE.C0[tT1.sfE.C0$P.Value < 0.05 & E4Coeff.C0 > 0, "GeneSymbol"]
E2.Genes.C0.2 = tT2.sfE.C0[tT2.sfE.C0$P.Value < 0.05 & E2Coeff.C0 < 0, "GeneSymbol"]
intersect(E4.Genes.C0.1, E2.Genes.C0.2)


# run Model 3
tT3_CERAD = run_model3("CERAD", mDataG, C, E, annot, "ROSMAP", "microglia") 
tT3_Braak = run_model3("Braak", mDataG, B, E, annot, "ROSMAP", "microglia")

nrow(tT3_CERAD[E4vsE3_P.Value < 0.05,])
nrow(tT3_Braak[E4vsE3_P.Value < 0.05,])
length(intersect(tT3_Braak[E4vsE3_P.Value < 0.05,]$GeneSymbol,tT3_CERAD[E4vsE3_P.Value < 0.05,]$GeneSymbol))

length(tT3_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
length(tT3_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)

length(tT3_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC > 0,]$GeneSymbol)
length(tT3_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC < 0,]$GeneSymbol)

length(tT3_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
length(tT3_Braak[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)


# run Model 4
tT4_CERAD = run_model4("CERAD", mDataG, C, E, annot, "ROSMAP", "microglia") 

nrow(tT4_CERAD[E4vsE3_P.Value < 0.05,])

length(tT4_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC > 0,]$GeneSymbol)
length(tT4_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_logFC < 0,]$GeneSymbol)

length(tT4_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC > 0,]$GeneSymbol)
length(tT4_CERAD[E2vsE3_P.Value < 0.05 & E2vsE3_logFC < 0,]$GeneSymbol)


# run Model 5
tT5_CERAD = run_model5("CERAD", mDataG, C, E, annot, "ROSMAP", "microglia") 

nrow(tT5_CERAD[C1E4vsC1E3_P.Value < 0.05,])
nrow(tT5_CERAD[C2E4vsC2E3_P.Value < 0.05,])

# prepare data for down stream analysis 
tT2_CERAD2 = tT2_CERAD
tT3_CERAD2 = tT3_CERAD
colnames(tT3_CERAD2) = gsub("E4vsE3", "C3E4vsC3E3",colnames(tT3_CERAD2))
colnames(tT3_CERAD2) = gsub("E2vsE3", "C3E2vsC3E3",colnames(tT3_CERAD2))
colnames(tT2_CERAD2) = gsub("E4vsE3", "C0E4vsC0E3",colnames(tT2_CERAD2))
colnames(tT2_CERAD2) = gsub("E2vsE3", "C0E2vsC0E3",colnames(tT2_CERAD2))

# pepare data for Figure 3
fig3 = tT2_CERAD2[tT5_CERAD, on=c("GeneSymbol", "EnsemblID")]
fig3 = fig3[tT3_CERAD2, on=c("GeneSymbol", "EnsemblID")]
```


# Z-Score Analysis and Spectral Clustering

```{r}
# z-score for C0 subjects only
cSubject = 0

# remove the 3 low-expressed genes: ENSG00000124091, ENSG00000252798, ENSG00000204345
mDataG1 = mDataG[!rownames(mDataG) %in% c("ENSG00000124091","ENSG00000252798","ENSG00000204345"),]

# select only the C0 subjects and remove the subject with APOE NA
mDataG1 = mDataG1[, which(C == cSubject)]
E.C0 = E[which(C == cSubject)]

means = rowMeans(mDataG1)
sds = apply(mDataG1, 1, sd)
zscore = function(x) return((x - means)/sds)
mDataZscore = apply(mDataG1, 2, zscore)

# check should be TRUE
cat("Is Z-score correct? ", all(mDataZscore[99,] == (mDataG1[99,] - means[99])/sds[99]), "\n")
```


# Average Z-Score

```{r}
# average z-score
z = data.table(Genes = annot$GeneSymbol[match(rownames(mDataZscore), annot$EnsemblID)],
               E2 = rowMeans(mDataZscore[, E.C0 == "E2"]),
               E3 = rowMeans(mDataZscore[, E.C0 == "E3"]),
               E4 = rowMeans(mDataZscore[, E.C0 == "E4"]))
```


# Spectral Clustering

```{r}
# spectral clustering
# microglia C0: k = 4; c = 2
k = 4
c = 2

set.seed(9)

zMtx = as.matrix(z[, c("E2", "E3", "E4")])
dissMatrix = dist(zMtx)
rownames(zMtx) = z$Genes
# compute similarity matrix
simMatrix = 1-as.matrix(dissMatrix)/max(dissMatrix)
 
# uses code from the Similarity Network Fusion Paper
# where kVal is the number of clusters you would like to partition
clust = SNFtool::spectralClustering(simMatrix, K = k) 
clustLab = as.factor(clust)

table(clustLab)


# Table S3
# detailed list of microglia-APOE cluster genes in ROSMAP DLPFC and MSBB STG CERAD 0 subjects
TableS3 = as.data.frame(cbind(clustLab, zMtx))
setDT(TableS3)
TableS3[, GeneSymbol := z$Genes]
TableS3 = annot2[GeneSymbol %in% TableS3$GeneSymbol][TableS3, on = .(GeneSymbol)]

colnames(TableS3) = c("EnsemblID", "GeneSymbol", "Description", "Cluster", "ROSMAP_E2AveZ", "ROSMAP_E3AveZ", "ROSMAP_E4AveZ")
TableS3 = TableS3[Cluster == 2,]
TableS3 = TableS3[,Cluster := NULL]
TableS3 = TableS3[order(GeneSymbol), ]
fwrite(TableS3, "TableS3.csv", row.names = F)
```


# Figure 2A and 2B
* Figure 2A is the heatmap for spectral clustering.
* Figure 2B is the individual z-score heatmap (C0 subjects).
* Column cluster (subjects) based on E2, E3, E4 group separately.
* Row cluster (genes) based on dendrogram.

```{r}
# prepare data
C2.genes = TableS3$GeneSymbol
E.C0 = factor(E[C == "0"], levels = c("E2", "E3", "E4"))
rownames(mDataZscore) = annot$GeneSymbol[match(rownames(mDataZscore), annot$EnsemblID)]
mDataZscore.c2 = mDataZscore[rownames(mDataZscore) %in% C2.genes,]
mDataZscore.c2H = apply(mDataZscore.c2, 2, function(x) return(ifelse(x > 4, 4, x)))
mDataZscore.c2H = apply(mDataZscore.c2H, 2, function(x) return(ifelse(x < -4, -4, x)))


# color function
colFun1 = colorRamp2(c(-1, -0.5, 0, 0.5, 1), c("#4575b4", "#74add1", "white", "#f46d43", "#d73027"))
colFun2 = colorRamp2(c(-4, -2, 0, 2, 4), c("#0000FFFF", "#7C50FDFF", "#EEEEEEFF", "#FF6545FF", "#FF0000FF"))


# Figure 2B
# individual z-score heatmap for ROSMAP C0 subjects
topAnnot = HeatmapAnnotation(z = anno_barplot(colMeans(mDataZscore.c2H), smooth = TRUE, axis_param = list(gp = gpar(fontsize = 5))), annotation_height = unit(0.7, "cm"))
h = Heatmap(mDataZscore.c2H, 
            name = "z-score",
            col = colFun2,
            show_row_names = FALSE,
            show_column_names = FALSE,
            cluster_rows = TRUE,
            show_heatmap_legend = FALSE,
            column_split = factor(E.C0, levels = c("E2", "E3", "E4")),
            cluster_column_slices = FALSE,
            column_title_gp =  gpar(col = "black", fontsize = 15),
            show_column_dend = TRUE,
            show_row_dend = TRUE,
            border = "#AAAAAA",
            row_dend_side = "right",
            height = unit(12.9, "cm"),
            column_dend_height = unit(0.7, "cm"),
            row_dend_width = unit(0.7, "cm"),
            width = unit(18, "cm"),
            top_annotation = topAnnot
)

panel_fun = function(index, nm) {
  pushViewport(viewport())
  grid.rect(gp = gpar(fill = "white", col = "white"))
  grid.lines(c(0, 1, 1, 0), c(0, 0, 1, 1), gp = gpar(col = "#AAAAAA"), default.units = "npc")
  draw(h, newpage = FALSE)
  popViewport()
}


# box for Figure 2B
layer_fun = function(j, i, x, y, width, height, fill) {
  v = pindex(zMtx[order(clustLab),], i, j)
  if(i %in% 105:251) {
    grid.rect(gp = gpar(lwd = 2, col = "black"))
  }
}

anno = anno_zoom(align_to = 105:251, which = "row", panel_fun = panel_fun, 
                 width = unit(21.5, "cm"), 
                 gap = unit(5, "cm"),
                 size = unit(15.8, "cm"),
                 link_width = unit(2, "cm"),
                 link_height = unit(5, "cm"),
                 link_gp = gpar(fill = "white", col = "#AAAAAA"),
                 internal_line = FALSE)


# legend
lgd.h = Legend(col_fun = colFun1, title = "A: Ave z-score", border = "black", 
               legend_width = unit(9.2, "cm"), 
               direction = "horizontal")

lgd.have = Legend(col_fun = colFun2, title = "B: z-score", border = "black", 
                  legend_width = unit(9.2, "cm"), 
                  direction = "horizontal")

pd = packLegend(lgd.h, lgd.have, 
                column_gap = unit(1, "cm"),
                direction = "horizontal")


# Figure 2A
# heatmap for average z-scores
pdf("../TablesFigures/Figure2AB.pdf", width = 11)
Heatmap(zMtx[order(clustLab),], 
        col = colFun1,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        show_row_names = FALSE,
        show_column_names = TRUE,
        right_annotation = rowAnnotation(foo = anno),
        row_split = sort(clustLab), 
        column_names_side = "bottom",
        layer_fun = layer_fun,
        show_heatmap_legend = FALSE,
        width = unit(5, "cm"),
        column_names_rot = 0,
        column_names_centered = T,
        row_title_gp = gpar(fontsize = 11)
)
draw(pd, x = unit(18, "cm"), y = unit(0.95, "cm"))
grid.text("A", x = unit(0.5, "cm"), y = unit(17.2, "cm"), gp = gpar(fontsize=20))
grid.text("B", x = unit(8.5, "cm"), y = unit(17.2, "cm"), gp = gpar(fontsize=20))
dev.off()
```


# Figure 2C

Violin plots where each dot represents a gene.

```{r}
G61 = intersect(C2.genes, tT_CERAD[E4vsE3_P.Value < 0.05 & E4vsE3_P.Value > 0, GeneSymbol])

mDataZscore.c2.e2 = mDataZscore.c2[,E.C0 == "E2"]
mDataZscore.c2.e3 = mDataZscore.c2[,E.C0 == "E3"]
mDataZscore.c2.e4 = mDataZscore.c2[,E.C0 == "E4"]

my.dat = list(`2` = rowMeans(mDataZscore.c2.e2), `3` = rowMeans(mDataZscore.c2.e3), `4` = rowMeans(mDataZscore.c2.e4))
my.df = reshape2::melt(my.dat)
my.df$L1 = as.numeric(my.df$L1)
colnames(my.df) = c("z-score", "APOE")
my.df$Genes = unlist(lapply(my.dat, function(x) names(x)))

ggthemr('greyscale')
colnames(my.df) = c("z-score", "APOE.num", "Genes")
my.df$APOE = as.factor(my.df$APOE.num)
  
to_swap = c("#62bba5", # E4
            "#785d37", # E3
            "#ffb84d") # E2
  
# violin plot
fig2c.1 = ggplot(my.df, aes(x=APOE.num, y=`z-score`, group=APOE, fill = APOE)) + 
  geom_violin(trim=FALSE) +
  scale_fill_manual(values = rev(to_swap)) +
  theme(legend.position = "none") +
  ylab("Ave z-score") +
  xlab("APOE") + 
  # plot the points
  geom_point(size=1, 
             colour="black", 
             alpha = 0.2,
             position=position_jitter(width=0.025)) +
  # plot the lines 
  geom_line(aes(group=Genes), 
            alpha = 0.1,
            color = "black")
  
# box plot
fig2c.2 = ggplot(data = my.df, aes(x=APOE, y=`z-score`, group=APOE, fill = APOE)) +
  geom_boxplot(outlier.colour="black", outlier.shape=16,
               outlier.size=2, notch=FALSE) +
  scale_fill_manual(values = rev(to_swap)) +
  ylab("Ave z-score")

fig2c = arrangeGrob(fig2c.1, fig2c.2, nrow=1, top = textGrob("ROSMAP",gp=gpar(fontsize=15,font=1)))
ggsave("../TablesFigures/Figure2C.pdf", fig2c, width = 5, height = 4)
```


# Figure 3B

E4 vs. E3 average Z-score line plot from C0 to C3.

```{r}
# remove the 3 low-expressed genes: ENSG00000124091, ENSG00000252798, ENSG00000204345
mDataG1 = mDataG[!rownames(mDataG) %in% c("ENSG00000124091","ENSG00000252798","ENSG00000204345"),]

# remove subjects which have no APOE or CERAD record
sel = (!is.na(E)) & (!is.na(C))
mDataG1 = mDataG1[, sel]

means = rowMeans(mDataG1)
sds = apply(mDataG1, 1, sd)
zscore = function(x) return((x - means)/sds)
mDataZscore = apply(mDataG1, 2, zscore)
  
# check should be TRUE
cat("Is Z-score correct? ", all(mDataZscore[99,] == (mDataG1[99,] - means[99])/sds[99]), "\n")

EC = paste(paste0("C", C[sel]), E[sel], sep = ":")
# average z-score
aveZ = data.table(Genes = annot$GeneSymbol[match(rownames(mDataZscore), annot$EnsemblID)],
                        
                        C0E2 = rowMeans(mDataZscore[, EC == "C0:E2"]),
                        C0E3 = rowMeans(mDataZscore[, EC == "C0:E3"]),
                        C0E4 = rowMeans(mDataZscore[, EC == "C0:E4"]),
                        
                        C1E2 = rowMeans(mDataZscore[, EC == "C1:E2"]),
                        C1E3 = rowMeans(mDataZscore[, EC == "C1:E3"]),
                        C1E4 = rowMeans(mDataZscore[, EC == "C1:E4"]),
                        
                        C2E2 = rowMeans(mDataZscore[, EC == "C2:E2"]),
                        C2E3 = rowMeans(mDataZscore[, EC == "C2:E3"]),
                        C2E4 = rowMeans(mDataZscore[, EC == "C2:E4"]),
                        
                        C3E2 = rowMeans(mDataZscore[, EC == "C3:E2"]),
                        C3E3 = rowMeans(mDataZscore[, EC == "C3:E3"]),
                        C3E4 = rowMeans(mDataZscore[, EC == "C3:E4"]))
 

# set up
ggthemr("fresh")
aveZ2 = melt(aveZ, id.vars = c("Genes"))
colnames(aveZ2) = c("Genes", "EC", "aveZ") 

to_swap = c("#62bba5", # E4
            "#785d37", # E3
            "#ffb84d") # E2

aveZ2[, CERAD := gsub("E[234]$", "", aveZ2$EC)]
aveZ2[, APOE := gsub("C[0123]", "", aveZ2$EC)]

pdf("../TablesFigures/Figure3A-147.pdf", width = 10, height = 5)
ggplot(aveZ2[Genes %in% C2.genes], aes(x = CERAD, y = aveZ, color = APOE)) + 
  geom_point() +
  geom_line(aes(group = factor(Genes)),
            color = "black",
            alpha = 0.1) +
  ylab("AveZ") +
  xlab("CERAD NP score") +
  facet_grid(cols = vars(APOE)) +
  scale_color_manual(values = rev(to_swap))
dev.off()

pdf("../TablesFigures/Figure3A.pdf", width = 10, height = 5)
fig3a = ggplot(aveZ2[Genes %in% G61], aes(x = CERAD, y = aveZ, color = APOE)) + 
  geom_point() +
  geom_line(aes(group = factor(Genes)),
            color = "black",
            alpha = 0.1) +
  ylab("AveZ") +
  xlab("CERAD NP score") +
  facet_grid(cols = vars(APOE)) +
  scale_color_manual(values = rev(to_swap))
fig3a
dev.off()
```


# Figure 3C

Correlation matrix comparing the APOE4 versus APOE3 change.

```{r}
fig3.1 = fig3[GeneSymbol %in% G61, .(C0E4vsC0E3_logFC, C1E4vsC1E3_logFC, C2E4vsC2E3_logFC, C3E4vsC3E3_logFC)]
fig3.2 = tidyr::gather(fig3.1, "CERAD", "logFC", 1:4) 


colnames(fig3.1) = c("C0", "C1", "C2", "C3")
ggthemr("fresh")
fig3.c = tidyr::gather(fig3.1, "CERAD", "logFC")
setDT(fig3.c)

axisLow = list("C0" = c("C1", "C2", "C3"),
               "C1" = c("C2", "C3"),
               "C2" = c("C3"))

axisUp = list("C1" = c("C0"),
              "C2" = c("C0", "C1"),
              "C3" = c("C0", "C1", "C2"))

color = list("C0" = "#f1c40f",
             "C1" = "#3498db",
             "C2" = "#2ecc71",
             "C3" = "#e74c3c")

p = list()
for(pCol in names(axisLow)){
  x = pCol
  for (y in axisLow[[pCol]]){
    label = paste0(x, y)
    p[["low"]][[label]] = ggplot(data = fig3.1, mapping = aes_string(x, y)) +
            geom_point(size = 0.5) +
            geom_abline(intercept = 0, slope = 1, color = "black", linetype = 2, alpha = 1.4) +
            xlim(-0.4, 1.2) +
            ylim(-0.4, 1.2) +
            theme(axis.line=element_line(color = "#786049"),
                  panel.border = element_rect(color = "#786049", fill = NA, size=1)) +
            ylab(y) +
            xlab(x) 
  }
}

for(pCol in names(axisUp)){
  x = pCol
  for (y in axisUp[[pCol]]){
    label = paste0(x, y)
    p[["up"]][[label]] = ggplot(fig3.c[CERAD %in% c(x, y), ], 
                aes(x = logFC, group = CERAD, fill = CERAD, color = CERAD)) +
         geom_density(alpha = 0.5) +
         ylab(y) +
         xlab(x) +
         xlim(-0.4, 1.2) +
         ylim(0, 4.5) +
         theme(panel.border = element_rect(color = "#786049", fill = NA, size=1)) + 
         scale_fill_manual(breaks = c("C0", "C1", "C2", "C3"), 
                           values = c("#f1c40f", "#3498db", "#2ecc71", "#e74c3c"))+ 
         scale_color_manual(breaks = c("C0", "C1", "C2", "C3"), 
                           values = c("#f1c40f", "#3498db", "#2ecc71", "#e74c3c")) + 
         theme(legend.position = "none")  
  }
}

for(pCol in c("C0", "C1", "C2", "C3")){
  x = pCol
  p[[x]] = ggplot() + 
           annotate("text", x = 4, y = 25, size=8, label = x, color = "#555555") + 
           theme(panel.border = element_rect(color = "#786049", fill = NA, size=1),
                 panel.grid.major = element_blank(),
                 axis.ticks = element_line(color = "white")
           ) + 
           scale_x_discrete(labels= " ") +
           scale_y_discrete(labels= " ") +
           xlab(x) +
           ylab(x) 
}

# legend 
pAllD = ggplot(fig3.c, 
                aes(x = logFC, group = CERAD, fill = CERAD, color = CERAD)) +
         geom_density(alpha = 0.5) +
         scale_fill_manual(breaks = c("C0", "C1", "C2", "C3"), 
                           values = c("#f1c40f", "#3498db", "#2ecc71", "#e74c3c"))+ 
         scale_color_manual(breaks = c("C0", "C1", "C2", "C3"), 
                           values = c("#f1c40f", "#3498db", "#2ecc71", "#e74c3c")) 
legend = cowplot::get_legend(pAllD)

fig3c = ggarrange(p[["C0"]],    p[["up"]][["C1C0"]],  p[["up"]][["C2C0"]],  p[["up"]][["C3C0"]],
          p[["low"]][["C0C1"]], p[["C1"]],            p[["up"]][["C2C1"]],  p[["up"]][["C3C1"]],
          p[["low"]][["C0C2"]], p[["low"]][["C1C2"]], p[["C2"]],            p[["up"]][["C3C2"]],
          p[["low"]][["C0C3"]], p[["low"]][["C1C3"]], p[["low"]][["C2C3"]], p[["C3"]],
          nrow = 4, ncol = 4, heights = c(3, 3, 3, 3), align = "hv")
fig3c2 = annotate_figure(fig3c,
                bottom = text_grob("log Fold Change E4 vs E3 in CERAD NP score", 
                                   size = 12, color = "#555555"),
                left = text_grob("log Fold Change E4 vs E3 in CERAD NP score", 
                                 size = 12, rot = 90, color = "#555555")
)
fig3c3 = cowplot::plot_grid(fig3c2, NULL, legend, rel_widths = c(4, 0.02, .3), nrow = 1)
ggsave("../TablesFigures/Figure3C.pdf", fig3c3, width = 9, height = 8)
```